ssz/utils.py

```diff
@@ -134,14 +139,19 @@ def hash_layer(child_layer: Sequence[bytes]) -> Tuple[Hash32, ...]:
     return parent_layer
 
 
-def merkleize(chunks: Sequence[Hash32], limit: int=None) -> Hash32:
+def merkleize(chunks: Sequence[Hash32], limit: int=None, cache=None) -> Hash32:
+    with_cache = cache is not None
+
     chunk_len = len(chunks)
 
     if limit is None:
         limit = chunk_len
 
     if limit == 0:
-        return ZERO_HASHES[0]
+        if cache is None:
+            return ZERO_HASHES[0]
+        else:
+            return ZERO_HASHES[0], cache
```

Comment:
Generally I think functions with inconsistent return signatures should be avoided.  They won't play nicely with `mypy` and they tend to require knowledge of the implementation or return value checking to interact with properly.

Reply:
Yes, totally agreed! This API needs to be refactored a lot and it's in my TODO list.
Workaround for now: returning `Union[Hash32, Tuple[Hash32, CacheObj]]`.